### ISSUE TITLE
HTTPS Redirects to HTTP after Project Delete

### ISSUE THREAD

**Issue by msymons** [DependencyTrack/dependency-track#277]
### Issue Type:
  - [X] defect report
  - [ ] enhancement request

### Current Behavior:
When accessing Dependency-Track via HTTPS (provided using nginx reverse proxy), all is OK until performing project delete.  After the deletion is (succesfully) complete, I receive a connection error:  DT server seems to have redirected me to use http:

`http://dependency-track.foo.bar.com/projects/`

instead of:

`https://dependency-track.foo.bar.com/projects/`

(There is no possibility of automatic HTTP -> HTTPS redirection in nginx as access to port 80 is blocked by a firewall).

### Steps to Reproduce (if defect):

- Set up Dependency-Track using HTTPS  using a reverse proxy.
- Ensure that there is no HTTP -> HTTPS redirection by checking that trying HTTP gives connection error.
- Navigate to a project (HTTPS)
- The URL is `https:/dependency-track.foo.bar.com/project/?projectuuid`
- Click "View Details" and then "Delete"
- After the deletion is complete DT server will redirect you to HTTP url and there will be connect error.

### Expected Behavior:
When using HTTPS, everything must be HTTPS.

### Environment:

  - Dependency-Track Version: 3.4.0
  - Distribution: [Executable WAR]
  - Browser:  Firefox v65.0

**Comment by stevespringett** [comment]
I don't have the infrastructure or resources to reproduce.  However, looking at the code, I do not see how this can possibly happen with a properly configured reverse proxy.

On a successful delete, this is the code that's called:
https://github.com/DependencyTrack/dependency-track/blob/master/src/main/webapp/project/functions.js#L500

No protocols are involved. It simply uses the current URL of the browser and navigates back to the projects page. I would recommend consulting the network team responsible for the proxy.

**Comment by msymons** [comment]
It seems that project deletion can take a while.  Perhaps as long as a few minutes (I was not timing it).  I wonder if nginx has a problem with that....

I will do as you suggest and investigate the proxy, and give feedback here if anything interesting turns up.

**Comment by msymons** [comment]
@stevespringett,  I should have been more specific in my original description and stated explicitly that I was using the Dependency-Track UI.

I did another test using the UI.  The browser sat for 3 minutes before displaying the connection error (HTTP instead of HTTPS).  As reported previously, the deletion itself was succesful.  The browser developer console did not help.. nothing was displayed for network traffic after I clicked the delete button.

But then I tried the "Swagger UI" for Firefox Extension.

* Project delete using the curl format  provided by the Swagger UI works great.  I tried it with a couple of uuid and curl reports that completion takes 3 seconds.  Note that curl can only connect to Dependency-Track via nginx proxy (just like the browser).
* "Execute" is also speedy (and also results in project deletion).  However,  the response is HTTP 204 and Swagger UI reports that this is undocumented.  Could that be the source of the problems experienced in the UI?

Response Headers:

```
cache-control: private, max-age=0, must-revalidate, no-cache
connection: keep-alive
content-security-policy: default-src 'self';script-src 'self' 'unsafe-inline';style-src 'self' 'unsafe-inline';img-src 'self';connect-src 'self';font-src 'self';object-src 'self';media-src 'self';frame-src 'self';child-src 'self';form-action 'self';frame-ancestors 'none';
date: Wed, 06 Feb 2019 10:59:43 GMT
server: nginx/1.14.0
x-frame-options: DENY
x-powered-by: Dependency-Track v3.4.0

```

**Comment by biasto** [comment]
We also have the same problem. After some investigation it seems that requests to `https://.../projects` are redirected to `http://` and requests to `/projects/` will stay on https. 

In Firefox Console, when I use `window.location.href="../projects/"`, it will return to the project overview on https.

**Comment by stevespringett** [comment]
@biasto I see three dots in your comment but the code only has two. Are you seeing requests being made to `https://.../projects` or `https://../projects`? Also, what is the value of `context-path` defined in one of the metas tag in HTML?

I'm trying to figure out if this is an issue with Dependency-Track or with FireFox since both reports have mentioned FireFox which has ~10% marketshare compared to Chromes 65% on the desktop/laptop which there have been no reports against.

**Comment by msymons** [comment]
I think that this is an issue with Dependency-Track (or something that could be improved in DT) and it just happens to be nginx that makes things appear confusing.

As I reported above,  testing with "swagger ui" and curl shows that project deletion is both completed and is speedy.

The problem is only in the UI.  If you delete a project you are doing it in the context of the project itself.  ... the URL that you are currently looking at has a uuid that will beocme invalid as soon as you perform the delete.  

When you execute the delete, the page does not switch to a valid URL.  Instead Firefox reports "waiting for..." and the page tab favicon has changed from DT favicon to a spinner.

So, if you sit back and wait then nginx will eventually do it's stuff and you end up with HTTPS -> HTTP and an error.

I think that DT should do the following following project delete, depending on situation:
* Project has just one version:  load the "Projects" screen
* Project has more than one version:  load the page for a version that is still remaining (eg top one)

**Comment by biasto** [comment]
I am very sorry to come back to this topic this late.

The three dots in my comment where meant as a replacement for the top-level-domain of the installed DT, e.g. https://web-dependency-track.container.example.com/projects. 

I wanted to point out that requests to `/projects` are being redirected to HTTP (Header: `Location: http://web-dependency-track.container.example.com/projects/`), whereas requests to `/projects/` are not redirected to HTTP. This happens in Firefox as well as in Chrome.

The context-path is: `<meta name="context-path" content="/">`, I hope this is what you meant.

My idea was, if the line https://github.com/DependencyTrack/dependency-track/blob/13aa3e1e8364501a7394e8faed207f31903b6a37/src/main/webapp/project/functions.js#L495 
would be changed to `window.location.href = "../projects/";` it could be working. I tested this in developer console in Firefox as well as in Chrome.

**Comment by stevespringett** [comment]
Since I'm not able to reproduce the issue, it would be good if someone could test with this fix and see if it resolves this issue or not. If not, there's nothing else I can think of that would help.

**Comment by msymons** [comment]
@stevespringett, I think that my comment (and suggested enhancement) from 3 April is still valid here.

In fact, it also would help if DT were to automatically reloaded the project screen after *adding* a new version.  That way, the new version would be immediately be available in the version-dropdown. Currently, one has to know that a manual page refresh is needed... otherwise it appears as if the version you just added errored in some way (which it did not).

**Comment by biasto** [comment]
Unfortunately I am not used to using github. I did not see that you already committed my suggestion.  I am sorry.

Today I realized that this problem no longer exists in version 3.6.1. After deleting one project I will be redirected to the project-overview on HTTPS.

**Comment by stevespringett** [comment]
Closing issue. Redirects are no longer used in new frontend (available soon in v3.8.0)
